### PR TITLE
Update for FIPS documentation

### PR DESCRIPTION
**Pull request by justsmth**
### Issues:
Resolves: CryptoAlg-2419

### Description of changes: 
Updating the documentation about our FIPS module.
* You can see the markdown rendered here: https://github.com/justsmth/aws-lc/blob/fips-doc-update/crypto/fipsmodule/FIPS.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

**Review by darylmartin100 — APPROVED** [review]

File: crypto/fipsmodule/FIPS.md [hunk 1]
```diff
@@ -1,175 +1,137 @@
...
+5. It adds a 64-byte, read-only array outside of the module to contain the known-good HMAC value.
 
-##### Integrity testing
+#### Integrity testing
 
 In order to actually implement the integrity test, a constructor function within the module calculates an HMAC from `module_start` to `module_end` using a fixed, all-zero key. It compares the result with the known-good value added (by the script) to the unhashed portion of the text segment. If they don't match, it calls `exit` in an infinite loop.
-
 Initially the known-good value will be incorrect. Another script (`inject_hash.go`) calculates the correct value from the assembled object and injects it back into the object.
 
-![build process](./intcheck2.png)
```
Comment (nebeid) [inline crypto/fipsmodule/FIPS.md:155]: Why we're removing this figure?

File: crypto/fipsmodule/FIPS.md [hunk 2]
```diff
@@ -1,175 +1,137 @@
...
-
-### Comparison with OpenSSL's method
```
Comment (nebeid) [inline crypto/fipsmodule/FIPS.md:157]: I usually found this section useful to come back to. Why is it being removed along with the figure?

File: crypto/fipsmodule/FIPS.md (hunk 1 above)
Comment (justsmth) [reply crypto/fipsmodule/FIPS.md:155]: I added back that image in the latest commit.

**Review by darylmartin100 — APPROVED** [review]

File: crypto/fipsmodule/FIPS.md [hunk 3]
```diff
@@ -1,175 +1,139 @@
...
+The textual assembly file is then processed by a “delocator” script to merge in assembly implementations of some primitives and to eliminate the remaining sources of relocations.
 
-##### Redirector functions
+#### Redirector functions
 
-The most obvious cause of relocations are out-calls from the module to non-cryptographic functions outside of the module. Most obviously these include `malloc`, `memcpy` and other libc functions, but also include calls to support code in BoringSSL such as functions for managing the error queue.
-
-Offsets to these functions cannot be known until the final link because only the linker sees the object files containing them. Thus calls to these functions are rewritten into an IP-relative jump to a redirector function. The redirector functions contain a single jump instruction to the real function and are placed outside of the module and are thus not hashed (see diagram).
-
-![module structure](./intcheck1.png)
```
Comment (nebeid) [inline crypto/fipsmodule/FIPS.md:123]: This figure was also removed.

File: crypto/fipsmodule/FIPS.md [hunk 4]
```diff
@@ -1,175 +1,139 @@
...
-1.  OpenSSL references read-write data directly, since it can know the offsets to it. BoringCrypto indirects these loads and stores.
-1.  OpenSSL doesn't run the power-on test until `FIPS_module_mode_set` is called. BoringCrypto does it in a constructor function. Failure of the test is non-fatal in OpenSSL, BoringCrypto will crash.
-1.  Since the contents of OpenSSL's module change between compilation and use, OpenSSL generates `fipscanister.o.sha1` to check that the compiled object doesn't change before linking. Since BoringCrypto's module is fixed after compilation (in the static case), the final integrity check is unbroken through the linking process.
-
-Some of the similarities are worth noting:
-
-1.  OpenSSL has all out-calls from the module indirecting via the PLT, which is equivalent to the redirector functions described above.
+The utility in `util/fipstools/break-hash.go` can be used to corrupt the FIPS module inside a binary and thus trigger a failure of the integrity test. Note that the binary must not be stripped, otherwise the utility will not be able to find the FIPS module.
+![build process](./intcheck2.png)
```
Comment (nebeid) [inline crypto/fipsmodule/FIPS.md:138]: Can there be a space left before the figure? It's too close to the text in the rendered output.

Comment (justsmth) [reply crypto/fipsmodule/FIPS.md:139]: Fixed in latest commit.

**Review by darylmartin100 — DISMISSED** [review]

**Review by darylmartin100 — APPROVED** [review]

**Review by nebeid — APPROVED** [review]

**Review by darylmartin100 — APPROVED** [review]

**Pull request merged**